src/server/frontend_wayland/wl_surface.cpp

```diff
@@ -366,49 +408,107 @@ void mf::WlSurface::commit(WlSurfaceState const& state)
-
-            buffer_size_ = new_buffer_size;
+            needs_buffer_submission = true;
         }
     }
     else
     {
         frame_callback_executor->spawn(std::move(executor_send_frame_callbacks));
     }
 
+    if (needs_buffer_submission)
+    {
+        auto const src_sample =
+            [&]() -> geom::RectangleD
+            {
+                auto const entire_buffer = geom::RectangleD{{0, 0}, geom::SizeD{current_buffer->size()}};
+                if (viewport && viewport.value().source)
+                {
+                    auto raw_source = viewport.value().source.value();
+                    /* Viewport coordinates are in buffer coordinates *after* applying transformation and scale.
+                     * That means this rectangle needs to be scaled. (And have buffer transform applied, when we support that)
+                     */
+                    geom::RectangleD source_in_buffer_coords{
+                        {raw_source.left().as_value() / inv_scale, raw_source.top().as_value() / inv_scale},
+                        raw_source.size / inv_scale
+                    };
+                    if (!entire_buffer.contains(source_in_buffer_coords))
+                    {
+                        throw wayland::ProtocolError{
+                            viewport.value().resource,
+                            wayland::Viewport::Error::out_of_buffer,
+                            "Source viewport (%f, %f), (%f × %f) is not entirely within buffer (%i × %i)",
+                            source_in_buffer_coords.left().as_value(),
+                            source_in_buffer_coords.top().as_value(),
+                            source_in_buffer_coords.size.width.as_value(),
+                            source_in_buffer_coords.size.height.as_value(),
+                            current_buffer->size().width.as_uint32_t(),
+                            current_buffer->size().height.as_uint32_t()
+                        };
+                    }
+                    return source_in_buffer_coords;
+                }
+                return entire_buffer;
+            }();
+        auto const logical_size =
+            [&]()
+            {
+                if (viewport)
+                {
+                    if (viewport.value().destination)
+                    {
+                        return viewport.value().destination.value();
+                    }
+                }
+                if (as_int_size_if_exact(src_sample.size))
+                {
+                    return as_int_size_if_exact(src_sample.size).value() * inv_scale;
+                }
+                else
+                {
+                    throw wayland::ProtocolError{
+                        viewport.value().resource,
+                        wayland::Viewport::Error::bad_size,
+                        "No wp_viewport destination set, and source size (%f × %f) is not integral", src_sample.size.width.as_value(), src_sample.size.height.as_value()};
```

Comment:
Same here. I would expect this to be thrown in the Viewporter protocol

Reply:
Likewise! We can only check this at `commit` time because we need to know the *`scale`* before we know what the source viewport actually is, and we only know the `scale` at `commit` time.